include/lbann/operators/operator.hpp

```diff
@@ -0,0 +1,209 @@
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef LBANN_OPERATORS_OPERATOR_HPP_INCLUDED
+#define LBANN_OPERATORS_OPERATOR_HPP_INCLUDED
+
+#include "lbann/base.hpp"
+#include "lbann/proto/datatype_helpers.hpp"
+#include "lbann/utils/cloneable.hpp"
+#include "lbann/utils/describable.hpp"
+#include "lbann/utils/exception.hpp"
+#include "lbann/utils/protobuf_serializable.hpp"
+#include "lbann/utils/tensor.hpp"
+#include "lbann/utils/typename.hpp"
+
+#include <operators.pb.h>
+
+#include <h2/meta/Core.hpp>
+#include <h2/meta/TypeList.hpp>
+
+#include <google/protobuf/message.h>
+
+#include <string>
+#include <vector>
+
+/** @brief A utility macro for easily adding ETI for operator builders
+ *  @note Must be called inside lbann namespace.
+ */
+#define LBANN_SINGLE_TYPE_OPERATOR_BUILDER_ETI(OPERATOR_NAME, T, D)            \
+  template std::unique_ptr<Operator<T, T, D>>                                  \
+    build_##OPERATOR_NAME##_operator<T, D>(lbann_data::Operator const&)
+
+namespace lbann {
+
+using supported_operator_data_type = h2::meta::TL<
+#ifdef LBANN_HAS_GPU_FP16
+  fp16,
+#endif
+#ifdef LBANN_HAS_HALF
+  cpu_fp16,
+#endif
+  float,
+  double>;
+
+/** @brief Neural network tensor operation.
+ *
+ *  An operator defines a mathematical function that that supports
+ *  both forward and possibly backward operations. In the forward
+ *  direction, it takes a vector of input tensors and produces a
+ *  vector of output tensors.  In the backward direction they
+ *  implement the differentiation of the forward operation, applying
+ *  the function to the operator's forward inputs and gradient with
+ *  respect to the outputs, to compute the gradient with respect to
+ *  the input.
+ *
+ *  Operators act as curried functions, they can have state that is
+ *  defined during construction but do not hold internal state.  A
+ *  operator should also be able to take objective function gradients
+ *  w.r.t. the outputs ("previous error signals") and compute the
+ *  objective function gradients w.r.t. the inputs ("error
+ *  signals"). This allows the model to perform automatic
+ *  differentiation.
+ *
+ *  Operators are specified for unique input and output data types.
+ */
+template <typename InputT, typename OutputT, El::Device D>
```

Comment:
We are making an assumption that all inputs have the same type and all outputs will have the same type. This is reasonable for most of our operations and it's best not to prematurely over-optimize, but we should anticipate cases where this may not hold:
- Indexing operations like gather or scatter, where one input is floating point and the other is integer.
- Operations where inputs are on different devices, e.g. a rotation layer where the image is on GPU and the rotation angle is on CPU.
- Operations without inputs (e.g. random distributions) or without outputs (e.g. outputs to objective functions or metrics).

Reply:
Yes for now we are making that assumption.  We will address these more complex operations in future designs. Possibly as manual layer, or a more complex operator interface.